core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java

```diff
@@ -179,37 +187,38 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
    *
    * @return array of lengths, in bytes, of each partition of the file (used by map output tracker).
    */
-  private long[] writePartitionedFile(File outputFile) throws IOException {
+  private long[] writePartitionedData(ShuffleMapOutputWriter mapOutputWriter) throws IOException {
     // Track location of the partition starts in the output file
     final long[] lengths = new long[numPartitions];
     if (partitionWriters == null) {
       // We were passed an empty iterator
       return lengths;
     }
 
-    final FileOutputStream out = new FileOutputStream(outputFile, true);
     final long writeStartTime = System.nanoTime();
-    boolean threwException = true;
     try {
       for (int i = 0; i < numPartitions; i++) {
         final File file = partitionWriterSegments[i].file();
-        if (file.exists()) {
-          final FileInputStream in = new FileInputStream(file);
-          boolean copyThrewException = true;
-          try {
-            lengths[i] = Utils.copyStream(in, out, false, transferToEnabled);
-            copyThrewException = false;
-          } finally {
-            Closeables.close(in, copyThrewException);
-          }
-          if (!file.delete()) {
-            logger.error("Unable to delete file for partition {}", i);
+        boolean copyThrewException = true;
+        // TODO: Enable transferTo
+        ShufflePartitionWriter writer = mapOutputWriter.getNextPartitionWriter();
+        try (OutputStream tempOutputStream = writer.openStream()) {
+          if (file.exists()) {
+            FileInputStream in = new FileInputStream(file);
+            try {
+              Utils.copyStream(in, tempOutputStream, false, false);
```

Comment:
So using `Utils.copyStream` actually assumes that the outputStream is a `FileOutputStream` in order to take advantage of the `transferTo` logic inside that function. The function also does its own bufferring with an array of a fixed side. I wonder if we should just have the `DefaultShuffleMapOutputWriter` just deal with generic `FileOutputStream`s and wrap them with buffered output writers if necessary? It would make the `DefaultShuffleMapOutputWriter` much cleaner too so we're not keeping track of so many layers of outputStreams

Reply:
Actually, on second thought, you probably just want to fork the code paths here into a "transferTo" block and a "outputstreams" block to allow more flexibility in the plugins to return whatever outputstreams they want.